Lib/test/test_syntax.py

```diff
@@ -148,6 +148,49 @@
 Traceback (most recent call last):
 SyntaxError: 'list' is an illegal expression for augmented assignment
 
+Invalid targets in `for` loops and `with` statements should also
+produce a specialized error message
+
+>>> for a() in b: pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'for' target
+
+>>> for (a, b()) in b: pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'for' target
+
+>>> for [a, b()] in b: pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'for' target
+
+>>> for (*a, b, c+1) in b: pass
+Traceback (most recent call last):
+SyntaxError: 'operator' is an illegal 'for' target
+
+>>> for (x, *(y, z.d())) in b: pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'for' target
+
+>>> with a as b(): pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'with' target
+
+>>> with a as (b, c()): pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'with' target
+
+>>> with a as [b, c()]: pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'with' target
+
+>>> with a as (*b, c, d+1): pass
+Traceback (most recent call last):
+SyntaxError: 'operator' is an illegal 'with' target
+
+>>> with a as (x, *(y, z.d())): pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'with' target
+
```

Comment:
These new tests will fail with the oldparser:

```
./python -X oldparser -m test test_syntax
...
test test_syntax failed -- 10 of 108 doctests failed
test_syntax failed

== Tests result: FAILURE ==

1 test failed:
    test_syntax

Total duration: 187 ms
Tests result: FAILURE

```

Reply:
You need to update the old parser to match these errors as well as it was done for the assigmnet case

Reply:
Another option is triying to produce the same errors as the old parser did, i.e. for this code:

```
>>> for [a, b()] in b: pass
```

produce

```
SyntaxError: cannot assign to function call
```

instead of

```
SyntaxError: 'function call' is an illegal 'for' target
```

Reply:
Yup, I agree it's better that way. Done in https://github.com/python/cpython/pull/20106/commits/1b6cf0b327f024dc53ab548bbfb2bca55064c31c.